cmd/registry/controller/parser.go

```diff
@@ -78,39 +110,51 @@ func extendDependencyPattern(
 
 }
 
+// This function is used in case of receipt artifacts where the name of the target resource to create is not known
 func resourceNameFromGroupKey(
 	resourcePattern string,
 	groupKey string) (string, error) {
-	// Derives the resource name from the provided resourcePattern and dependencyName.
+	// Derives the resource name from the provided resourcePattern and groupKey.
 	// Example:
 	// 1) resourcePattern: projects/demo/locations/global/apis/-/versions/-/specs/-
-	//    dependencyName: projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/complexity
+	//    groupKey: projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/complexity
 	//    returns projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml
 	// 2) resourcePattern: projects/demo/locations/global/apis/petstore/versions/-/specs/-/artifacts/custom-artifact
-	//    dependencyName: projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/complexity
+	//    groupKey: projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/complexity
 	//    returns projects/demo/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/custom-artifact
 
-	// Replace apis/- pattern with the corresponding name and so on.
-	// dependency.GetApi() returns the full api name projects/demo/locations/global/apis/petstore
-	// We use stringsSplit()[-1] to extract only the API name
-	// apiPattern := regexp.MustCompile(`/apis/-`)
-	apiName := strings.Split(extractEntityName(groupKey, "api"), "/")
-	resourceName := strings.ReplaceAll(resourcePattern, "/apis/-",
-		fmt.Sprintf("/apis/%s", apiName[len(apiName)-1]))
-
-	versionName := strings.Split(extractEntityName(groupKey, "version"), "/")
-	resourceName = strings.ReplaceAll(resourceName, "/versions/-",
-		fmt.Sprintf("/versions/%s", versionName[len(versionName)-1]))
+	groupName, err := parsePattern(groupKey)
+	resourceName := resourcePattern
```

Comment:
This copy was confusing to me as a reader because it gives two names for the same value. Eventually I realized its purpose is to maintain the original parameter value so it can be used in the error message at the end of the function. 

Can this function accept resource name types as parameters instead of raw strings? That would eliminate the need for validation entirely, and would also give the function more tools to manipulate the names instead of using string replacement.

Reply:
I agree the current approach is not the best with string replacement operations. the resourceName is a very generic pattern, it can be an Api, Spec, Version or Artifact. Currently, the way the controller works, it doesn't identify the type of a resource or a resourcePattern (hence the use of generic resource interface). 

However, we can always parse the pattern and figure out what kind it is. I would like to revisit this in a separate PR. I am guessing, changing this logic will need some changes to the interfaces we have defined. Plus if we are making changes to `names` package, then. this logic will be different.

Reply:
Yeah, we should think about how to address this in the names package itself. Definitely not in the scope of this PR. I would like our names package to provide good enough abstractions that client code like this doesn't need to create wrappers.

Reply:
I will revisit this and take another pass at the `names` package.